File: dp/two-sets-ii/two-sets-ii.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

int dp[501][100000];
int ready[501][100000];
int MOD = (int) (1e9 + 7);

int backtrack(int i, int j){
    if(j == 0){
        return 1;
    }
    if(i < 0 || j < 0){
        return 0;
    }
    if(ready[i][j])
        return dp[i][j];

    dp[i][j] = backtrack(i - 1, j);
    if(j >= i){
        dp[i][j] += backtrack(i - 1, j - i);
    }
    dp[i][j] %= MOD;
    ready[i][j] = true;
    return dp[i][j];
}

int main(){
    int n;
    cin >> n;
    if(n%4 == 0 || (n+1)%4 == 0){
        int target = ((n+1)*n) / 4;
        int res = backtrack(n - 1, target);
        cout << res << endl;
    }
    else{
        cout << 0 << endl;
    }
}

File: dp/two-sets-ii/two-sets-ii.py
from sys import *
memo = {}
MOD = int(1e9 + 7)
def backtrack(i,j):
    if (i,j) in memo:
        return memo[(i,j)]
    if j == 0:
        return 1
    if i <= 0 or j < 0:
        return 0
    memo[(i,j)] = backtrack(i - 1, j) + backtrack(i - 1, j - i)
    return memo[(i,j)]
def solve(n):  
    if not (n%4 == 0 or (n+1)%4 == 0):
        print(0)
        return
    else:
        target = n*(n+1)/4
        print((backtrack(n, target) * pow(2, MOD - 2, MOD))  % MOD)
        return
if __name__ == "__main__":
    n = int(stdin.readline().strip())
    solve(n)
